### ISSUE TITLE
False positive, possibly related to type inference and lambdas

### ISSUE THREAD

**Issue by kelloggm** [typetools/checker-framework#3036]
Reduced from one of the errors reported by @msridhar in  [#3001-comment](https://github.com/typetools/checker-framework/issues/3001#issuecomment-567616512). An error is issued by most (all?) type systems, including the Signature Checker and the Object Construction Checker. I'm looking into the root cause.

Test case:

```
import java.util.List;
import java.util.Set;
import java.util.Map;
import java.util.HashMap;
import java.util.stream.Collectors;

public class LambdaInStream {

    public Set<MyInnerClass> getDsData() {
        return null;
    }

    public static class MyInnerClass {
        public int getKeyTag() {
            return 5;
        }

        public String getDigest() {
            return null;
        }
    }

    private void write() {
        List<Map<String, ?>> dsData = 
                        getDsData()
                        .stream()
                        .map(dsData1 ->
                                        ImmutableMap.of(
                                                "keyTag", dsData1.getKeyTag(),
                                                "digest", dsData1.getDigest()))
                        .collect(Collectors.toList());
    }

    public static class ImmutableMap<K, V> extends HashMap<K, V> {
        public static <K, V> ImmutableMap<K, V> of(K k1, V v1, K k2, V v2) {
            return null;
        }
    }
}
```

Error:
```
> javac -processor signature framework/tests/all-systems/java8inference/LambdaInStream.java
framework/tests/all-systems/java8inference/LambdaInStream.java:30: error: [return.type.incompatible] incompatible types in return.
                                        ImmutableMap.of(
                                                       ^
  type of expression: @SignatureUnknown ImmutableMap<@Identifier String, @SignatureUnknown Serializable & @SignatureUnknown Comparable<?[ extends @SignatureUnknown Serializable & @SignatureUnknown Comparable<?[ extends @SignatureUnknown Object super @SignatureUnknown Void]> super @SignatureUnknown Void]>>
  method return type: @SignatureUnknown ImmutableMap<@SignatureUnknown String, @SignatureUnknown Serializable & @SignatureUnknown Comparable<?[ extends @SignatureUnknown Serializable & @SignatureUnknown Comparable<?[ extends @SignatureUnknown Object super @SignatureBottom Void]> super @SignatureBottom Void]>>
1 error
```
